### PR TITLE
Send logs over HTTP

### PR DESCRIPTION
**Pull request by achntrl**
### What does this PR do?

The PR implements a HTTP client to send logs to our HTTP endpoints. For now, it's not the default behavior, and the HTTP client is hidden behind the new `logs_config.use_http` option

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Follow-up [PR](https://github.com/DataDog/datadog-agent/pull/3594)

### CODE REVIEW

File: pkg/logs/pipeline/pipeline.go [hunk 1]
```diff
@@ -17,36 +19,37 @@ import (
...
-	}
-
+func NewPipeline(outputChan chan *message.Message, processingRules []*config.ProcessingRule, endpoints *config.Endpoints, destinationsContext *tcp.DestinationsContext) *Pipeline {
 	// initialize the sender
-	destinations := client.NewDestinations(main, additionals)
+	destinations := client.NewDestinations(endpoints, destinationsContext)
 	senderChan := make(chan *message.Message, config.ChanSize)
-	sender := sender.NewSender(senderChan, outputChan, destinations)
+	var newsender sender.Sender
+	if coreConfig.Datadog.GetBool("logs_config.use_http") {
```
Comment (ajacquemot) [inline pkg/logs/pipeline/pipeline.go]: nits: I think we could merge sender and encoder constructions inside the same block

File: pkg/logs/pipeline/pipeline.go [hunk 2]
```diff
@@ -17,36 +19,37 @@ import (
...
+	}
 
 	// initialize the input chan
 	inputChan := make(chan *message.Message, config.ChanSize)
 
 	// initialize the processor
-	encoder := processor.NewEncoder(endpoints.Main.UseProto)
+	var encoder processor.Encoder
+	if coreConfig.Datadog.GetBool("logs_config.use_http") {
+		encoder = processor.NewHTTPEncoder()
```
Comment (ajacquemot) [inline pkg/logs/pipeline/pipeline.go]: I don't think we should processing and network are different concerns so the encoder used `with` (not `for`) HTTP should not be called HTTPEncoder but JSONEncoder instead.

File: pkg/logs/pipeline/pipeline.go [hunk 3]
```diff
@@ -17,36 +19,37 @@ import (
...
+	} else {
+		encoder = processor.NewTCPEncoder(endpoints.Main.UseProto)
```
Comment (ajacquemot) [inline pkg/logs/pipeline/pipeline.go]: For this one, we should have ProtoEncoder or RawEncoder depending on use_proto config flag.

File: pkg/logs/processor/encoder.go [hunk 4]
```diff
@@ -28,14 +31,21 @@ var rawEncoder raw
 // Proto is an encoder implementation that writes messages as protocol buffers.
 var protoEncoder proto
 
-// NewEncoder returns an encoder.
-func NewEncoder(useProto bool) Encoder {
+// jsonEncoder is an encoder implementation that writes messages as JSON
+var aJsonEncoder jsonEncoder
```
Comment (ajacquemot) [inline pkg/logs/processor/encoder.go]: Let's be consistent
```suggestion
var jsonEncoder json
```

File: pkg/logs/sender/config.go [hunk 5]
```diff
@@ -11,11 +11,12 @@ import (
...
 
 	"github.com/DataDog/datadog-agent/pkg/config"
-	"github.com/DataDog/datadog-agent/pkg/logs/client"
+	logsConfig "github.com/DataDog/datadog-agent/pkg/logs/config"
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 )
 
-const endpointPrefix = "agent-intake.logs."
+const tcpEndpointPrefix = "agent-intake.logs."
+const httpEndpointPrefix = "https://http-intake.logs."
```
Comment (ajacquemot) [inline pkg/logs/sender/config.go]: what if we want to use `http` instead of `https` ?

File: pkg/logs/sender/sender.go [hunk 6]
```diff
@@ -67,7 +76,7 @@ func (s *Sender) send(payload *message.Message) {
 				break
 			}
 			switch err.(type) {
-			case *client.FramingError:
+			case *tcp.FramingError:
```
Comment (ajacquemot) [inline pkg/logs/sender/sender.go]: `sender` should be agnostic from the inner transport so we should not have a reference to the `tcp` package.

File: pkg/logs/sender/sender.go [hunk 7]
```diff
@@ -7,55 +7,64 @@ package sender
...
 	"github.com/DataDog/datadog-agent/pkg/logs/client"
+	"github.com/DataDog/datadog-agent/pkg/logs/client/tcp"
 	"github.com/DataDog/datadog-agent/pkg/logs/message"
 	"github.com/DataDog/datadog-agent/pkg/logs/metrics"
 )
 
-// Sender is responsible for sending logs to different destinations.
-type Sender struct {
+// Sender sends logs to different destination and reports to auditor
+type Sender interface {
```
Comment (ajacquemot) [inline pkg/logs/sender/sender.go]: it feels like there is an implementation per protocol so I would put the tcp logic under a `tcp` package and the `http` logic under a `http` package

File: pkg/logs/sender/sender.go [hunk 8]
```diff
@@ -89,3 +98,117 @@ func (s *Sender) send(payload *message.Message) {
...
+
+// NewBatchSender returns an new BatchSender.
+func NewBatchSender(inputChan, outputChan chan *message.Message, destinations *client.Destinations) *BatchSender {
+	return &BatchSender{
+		inputChan:     inputChan,
+		outputChan:    outputChan,
+		destinations:  destinations,
+		done:          make(chan struct{}),
+		batchTimeout:  5 * time.Second,               // TODO Change this to paramter
+		messageBuffer: NewMessageBuffer(20, 2000000), // TODO change this to paramter
```
Comment (ajacquemot) [inline pkg/logs/sender/sender.go]: 2MB is too high I think, I would lower it down to 1MB

File: pkg/logs/sender/sender.go [hunk 9]
```diff
@@ -89,3 +98,117 @@ func (s *Sender) send(payload *message.Message) {
...
+
+	for {
+		select {
+		case payload, isOpen := <-b.inputChan:
+			if !isOpen {
+				// inputChan has been closed, no more payload are expected
+				// TODO remember to flush your buffer!!
+				return
+			}
+			sendNow, appendAgain := b.messageBuffer.tryAppendToBuffer(payload)
```
Comment (ajacquemot) [inline pkg/logs/sender/sender.go]: I think append to buffer should not return any logic to know when to send, when to append, this logic should be done inside this method.

File: pkg/logs/sender/message_buffer.go [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+	"github.com/DataDog/datadog-agent/pkg/logs/message"
+)
+
+// MessageBuffer accumulates lines in buffer escaping all '\n'
+// and accumulates the total number of bytes of all lines in line representation (line + '\n') in rawDataLen
+type MessageBuffer struct {
+	maxBatchCount  int
+	maxRequestSize int
+	messageBuffer  []*message.Message
+	byteBuffer     []byte
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]: We should use a `bytes.Buffer`, see https://golang.org/pkg/bytes/#Buffer

File: pkg/logs/sender/message_buffer.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+
+package sender
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/message"
+)
+
+// MessageBuffer accumulates lines in buffer escaping all '\n'
+// and accumulates the total number of bytes of all lines in line representation (line + '\n') in rawDataLen
+type MessageBuffer struct {
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]: There is no public method for this class.

File: pkg/logs/sender/message_buffer.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+	maxBatchCount  int
+	maxRequestSize int
+	messageBuffer  []*message.Message
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]: there is no need to store the message buffer just to track it's content length, we could just use a counter instead.

**Review by ajacquemot — CHANGES_REQUESTED** [review]
Left a bunch of comments, there is still a bunch of work to do.

File: pkg/logs/sender/config.go [hunk 13]
```diff
@@ -71,25 +78,37 @@ func BuildEndpoints() (*client.Endpoints, error) {
...
 		additionals[i].UseSSL = useSSL
-		additionals[i].UseProto = useProto
 		additionals[i].ProxyAddress = proxyAddress
 	}
 
-	return client.NewEndpoints(main, additionals), nil
+	return logsConfig.NewEndpoints(main, additionals, useProto, false), nil
+}
+
+func buildHTTPEndpoints() (*logsConfig.Endpoints, error) {
```
Comment (ajacquemot) [inline pkg/logs/sender/config.go]: @ajacquemot add support for `logs_config.logs_dd_url`

Comment (ajacquemot) [reply pkg/logs/sender/config.go]: @ajacquemot fix `useSSL`

File: pkg/logs/client/http/destination.go [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/DataDog/datadog-agent/pkg/logs/config"
+	"github.com/DataDog/datadog-agent/pkg/util"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+const contentType = "application/json"
+
+// Destination sends a payload over HTTP.
+type Destination struct {
+	endpoint            config.Endpoint
```
Comment (ajacquemot) [inline pkg/logs/client/http/destination.go]: @ajacquemot remove this, not needed

File: pkg/logs/sender/message_buffer.go [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+func (mb *MessageBuffer) tryAppendToBuffer(m *message.Message) bool {
+	if (len(mb.messageBuffer) < cap(mb.messageBuffer)) && mb.hasSpaceInByteBuffer(m.Content) {
+		// fits. append the message
+		mb.messageBuffer = append(mb.messageBuffer, m)
+		mb.appendByteBuffer(m.Content)
+		return true
+	}
+
+	// doesn't fit, which can only be caused by not enough space in byteBiffer
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]:
```suggestion
	// doesn't fit, which can only be caused by not enough space in byteBuffer
```

File: pkg/logs/sender/message_buffer.go [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+	mb.messageBuffer = mb.messageBuffer[:0]
+	mb.byteBuffer = mb.byteBuffer[:1] // keep the first byte, it's used for : '['
+}
+
+func (mb *MessageBuffer) hasSpaceInByteBuffer(content []byte) bool {
+	return len(mb.byteBuffer)+len(content)+1 < cap(mb.byteBuffer)
+}
+
+func (mb *MessageBuffer) appendByteBuffer(content []byte) {
+	// increase the slice length, TODO can optimized this by not using append
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]: and use a bytes.Buffer for example which has been designed to append bytes in an optimized way ? ;)

File: pkg/logs/sender/message_buffer.go [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+package sender
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/logs/message"
+)
+
+// MessageBuffer accumulates lines in buffer escaping all '\n'
+// and accumulates the total number of bytes of all lines in line representation (line + '\n') in rawDataLen
+type MessageBuffer struct {
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]: This class has two responsibilities, buffering messages and building the payload and I feel like we should rename it as well as define a clear API.
For example:
```
type BatchPayloadBuilder interface {
  func TryAddMessage(msg message.Message) bool
  func Build() []byte
  func Clear()
  func GetMessages() [] message.Message
  func IsEmpty() bool
  func IsFull() bool
}
```

File: pkg/logs/sender/batch_sender.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+			b.sendBuffer()
+			flushTimer.Reset(b.batchTimeout)
+		}
+	}
+}
+
+// send keeps trying to send the message to the main destination until it succeeds
+// and try to send the message to the additional destinations only once.
+func (b *BatchSender) sendBuffer() {
+	if b.messageBuffer.isEmpty() {
```
Comment (ajacquemot) [inline pkg/logs/sender/batch_sender.go]: I feel like this check should be done in the callee

File: pkg/logs/sender/batch_sender.go [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+			// the timout expired, the content is ready to be sent
+			b.sendBuffer()
+			flushTimer.Reset(b.batchTimeout)
+		}
+	}
+}
+
+// send keeps trying to send the message to the main destination until it succeeds
+// and try to send the message to the additional destinations only once.
+func (b *BatchSender) sendBuffer() {
```
Comment (ajacquemot) [inline pkg/logs/sender/batch_sender.go]: this method is a duplicate, let's factorise it.

File: pkg/logs/client/http/destination.go [hunk 20]
```diff
@@ -32,18 +34,32 @@ func NewDestination(endpoint config.Endpoint) *Destination {
...
 			Timeout:   time.Second * 10,
 			Transport: util.CreateHTTPTransport(),
 		},
+		destinationsContext: destinationsContext,
 	}
 }
 
 // Send sends a payload over HTTP.
 func (d *Destination) Send(payload []byte) error {
-	resp, err := d.client.Post(d.url, contentType, strings.NewReader(string(payload)))
```
Comment (ajacquemot) [inline pkg/logs/client/http/destination.go]: Why can't we stick to this method, it looks cleaner IMO ? We don't do anything fancy on the request.

File: pkg/logs/client/http/destination.go [hunk 21]
```diff
@@ -32,18 +34,32 @@ func NewDestination(endpoint config.Endpoint) *Destination {
...
+
+	ctx := d.destinationsContext.Context()
+	req, err := http.NewRequest("POST", d.url, strings.NewReader(string(payload)))
```
Comment (ajacquemot) [inline pkg/logs/client/http/destination.go]: We need a bunch of constants here:
- POST
- ContentType
- ApplicationJSON
- ...

File: pkg/logs/sender/message_buffer.go [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// Clear removes all elements from the buffer.
+func (mb *MessageBuffer) Clear() {
+	mb.messageBuffer = mb.messageBuffer[:0]
+	mb.byteBuffer = mb.byteBuffer[:1] // keep the first byte, it's used for : '['
+}
+
+// GetPayload returns the concatanated messages in JSON encoded format.
+func (mb *MessageBuffer) GetPayload() []byte {
```
Comment (ajacquemot) [inline pkg/logs/sender/message_buffer.go]: We could also use `NJON` witch content type `text/plain` for better performances.

**Review by achntrl — COMMENTED** [review]
LGTM, I like how things are packaged 👍

**Review by ajacquemot — APPROVED** [review]

**Pull request merged**